py/formatfloat: Fix formatting of whole-numbered floats.

dpwe opened the pull request:
This is to fix issue https://github.com/micropython/micropython/issues/4212

Many bugs occur when printing float values containing integer values.  The successive-scaling approach in the original algorithm would quickly make these values depart from whole numbers.

This modification attempts to avoid scaling the original value for numbers in the range 1.0 to 1e10, and instead compares with other floats constructed only by multiplying whole numbers, so guaranteed to remain exact until the mantissa runs out.

I added a test that verifies consistency with CPython for a range of values.  In fact, CPython also struggles with some values (see, e.g. ```print(13 * .001)```), but at least we have agreement for the values in the test (on my machine).  

Signed-off-by: Dan Ellis <dan.ellis@gmail.com>

dpgeorge commented:
Thanks for the contribution, this looks really good!  So far my testing of this PR has shown that it does do a better job than before.  One test I used was micropython-ulab's tests suite and it fixed some printing issues there.

dpwe commented:
After looking at @goatchurchprime’s earlier proposed fix, I changed mine to use integer math for the whole-number part.  I also changed the test to only test whole-numbers (since that is the focus of the fix) and added a couple of specific cases.

dpgeorge commented:
There is something wrong with this branch, it now has 2 copies of the patch.  Can you please fix it, ideally rebasing on master?

dpwe commented:
I'm clumsy with git and couldn't figure out how to fix this without starting over.  Please see #8901 for the cleaned-up version of this change.
